smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/ResourceOperationInputOutputValidator.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.validation.validators;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.IdentifierBindingIndex;
+import software.amazon.smithy.model.knowledge.MemberPropertyIndex;
+import software.amazon.smithy.model.knowledge.OperationIndex;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.NestedPropertiesTrait;
+import software.amazon.smithy.model.traits.PropertyTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * Validates that resource are applied appropriately to resources.
+ */
+public final class ResourceOperationInputOutputValidator extends AbstractValidator {
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        return model.shapes(ResourceShape.class)
+                .filter(ResourceShape::hasProperties)
+                .flatMap(shape -> validateResource(model, shape).stream())
+                .collect(Collectors.toList());
+    }
+
+    private List<ValidationEvent> validateResource(Model model, ResourceShape resource) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        Set<String> propertiesInOperations = new TreeSet<>();
+        OperationIndex operationIndex = OperationIndex.of(model);
+        IdentifierBindingIndex identifierBindingIndex = IdentifierBindingIndex.of(model);
+        MemberPropertyIndex memberPropertyIndex = MemberPropertyIndex.of(model);
+
+        resource.getPut().flatMap(model::getShape).flatMap(Shape::asOperationShape).ifPresent(operation -> {
```

Comment:
Can this be refactored into some kind of abstracted method instead of repeating this multiple times?

Reply:
I also thought we were validating all instance operations too, no?

Reply:
Refactored and added `resource.getOperations()` as well.